### PR TITLE
Nonrandom stub and existence checks

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
A couple of changes in this PR, mostly small.

1) Altered all of the currently used stubs to return a truncated amount of actual data rather than random noise.
2) Added more file existence checks prior to file access.
3) Added improved type checking to functions.
4) Enforced PEP8 throughout.
5) Enforced Numpy docstring format throughout.

6) Also found a trailing bs4 soup reference in the currently unused waveforms call; fixed to reference appropriate field of lxml root.

### CODE REVIEW

**Comment by CodyCBakerPhD** [comment]
Still left to do in another couple PR's is...

A) Instead of defaulting certain parameters such as `fs` to hard-coded values, retrieve these from an `.XML` file.
B) Replace the use of `os.path` with `pathlib`.

C) Deal with functions like `read_spike_times` and `read_spike_clustering` with regards to our new usage of `SpikeInterface`, which contains all of this information. This could include extending some of the functions in this library that inter-rely on these calls to alternatively receiving a sorting extractor object assumed to contain the appropriate information.

How does that all sound?

File: to_nwb/neuroscope.py [hunk 1]
```diff
@@ -1,29 +1,39 @@
...
+        lxml.etree root object of the xml document.
+
+    """
     return et.parse(xml_filepath).getroot()
 
 
-def get_channel_groups(session_path=None, xml_filepath: str = None):
-    """Get the groups of channels that are recorded on each shank from the xml
-    file
+def get_channel_groups(session_path: str, xml_filepath: Optional[str]):
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
def get_channel_groups(session_path: str, xml_filepath: Optional[str] = None):
```

File: to_nwb/neuroscope.py [hunk 2]
```diff
@@ -32,23 +42,30 @@ def get_channel_groups(session_path=None, xml_filepath: str = None):
...
+        print("Warning: No .xml file found at the path location!"
+              "Unable to retrieve channel_groups.")
+        channel_groups = None
 
     return channel_groups
 
 
-def get_shank_channels(session_path=None, xml_filepath: str = None):
-    """Read the channels on the shanks in Neuroscope xml
+def get_shank_channels(session_path: str, xml_filepath: Optional[str]):
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
def get_shank_channels(session_path: str, xml_filepath: Optional[str] = None):
```

File: to_nwb/neuroscope.py [hunk 3]
```diff
@@ -64,17 +81,21 @@ def get_shank_channels(session_path=None, xml_filepath: str = None):
...
+              "Unable to retrieve shank_channels.")
+        shank_channels = None
 
     return shank_channels
 
 
-def get_lfp_sampling_rate(session_path=None, xml_filepath=None):
-    """Reads the LFP Sampling Rate from the xml parameter file of the
-    Neuroscope format
+def get_lfp_sampling_rate(session_path: str, xml_filepath: Optional[str]):
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
def get_lfp_sampling_rate(session_path: str, xml_filepath: Optional[str] = None):
```

File: to_nwb/neuroscope.py [hunk 4]
```diff
@@ -32,23 +42,30 @@ def get_channel_groups(session_path=None, xml_filepath: str = None):
...
-                      for group in root.find('anatomicalDescription').find('channelGroups').findall('group')]
+    if os.path.isfile(xml_filepath):
+        root = load_xml(xml_filepath)
+        channel_groups = [[int(channel.text)
+                          for channel in group.findall('channel')]
+                          for group in root.find('anatomicalDescription').find('channelGroups').findall('group')]
+    else:
+        print("Warning: No .xml file found at the path location!"
+              "Unable to retrieve channel_groups.")
+        channel_groups = None
```
Comment (bendichter) [inline to_nwb/neuroscope.py]: if there is no xml file, the function should fail

File: to_nwb/neuroscope.py [hunk 5]
```diff
@@ -64,17 +81,21 @@ def get_shank_channels(session_path=None, xml_filepath: str = None):
...
-                      for group in root.find('spikeDetection').find('channelGroups').findall('group')]
+    if os.path.isfile(xml_filepath):
+        root = load_xml(xml_filepath)
+        shank_channels = [[int(channel.text)
+                          for channel in group.find('channels')]
+                          for group in root.find('spikeDetection').find('channelGroups').findall('group')]
+    else:
+        print("Warning: No .xml file found at the path location!"
+              "Unable to retrieve shank_channels.")
+        shank_channels = None
```
Comment (bendichter) [inline to_nwb/neuroscope.py]: if there is no xml file, the function should fail

File: to_nwb/neuroscope.py [hunk 6]
```diff
@@ -90,15 +111,20 @@ def get_lfp_sampling_rate(session_path=None, xml_filepath=None):
...
 
-    root = load_xml(xml_filepath)
-    lfp_sampling_rate = float(root.find('fieldPotentials').find('lfpSamplingRate').text)
+    if os.path.isfile(xml_filepath):
+        root = load_xml(xml_filepath)
+        lfp_sampling_rate = float(root.find('fieldPotentials').find('lfpSamplingRate').text)
+    else:
+        print("Warning: No .xml file found at the path location!"
+              "Unable to retrieve lfp_sampling_rate.")
+        lfp_sampling_rate = None
```
Comment (bendichter) [inline to_nwb/neuroscope.py]: if there is no xml file, the function should fail

File: to_nwb/neuroscope.py [hunk 7]
```diff
@@ -249,12 +275,17 @@ def write_unit_series(nwbfile, session_path, shankn, fs=20000.):
...
-                          max_shanks=None):
-    """
+def write_electrode_table(nwbfile: NWBFile, session_path: str,
+                          electrode_positions: Optional[ArrayLike],
+                          impedances: Optional[ArrayLike],
+                          locations: Optional[ArrayLike],
+                          filterings: Optional[ArrayLike],
+                          custom_columns: Optional[List[dict]],
+                          max_shanks: Optional[int]):
+    """Write the electrode table to the NWBFile object.
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
def write_electrode_table(nwbfile: NWBFile, session_path: str,
                          electrode_positions: Optional[ArrayLike] = None,
                          impedances: Optional[ArrayLike] = None,
                          locations: Optional[ArrayLike] = None,
                          filterings: Optional[ArrayLike] = None,
                          custom_columns: Optional[List[dict]] = None,
                          max_shanks: Optional[int] = None):
    """Write the electrode table to the NWBFile object.
```

File: to_nwb/neuroscope.py [hunk 8]
```diff
@@ -318,57 +349,68 @@ def write_electrode_table(nwbfile, session_path, electrode_positions=None,
...
 
-def read_lfp(session_path, stub=False):
-    """
+def read_lfp(session_path: str, stub: bool = False):
+    """Read LFP data from Neuroscope eeg file.
 
     Parameters
     ----------
-    session_path: str
+    session_path: PathType
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
    session_path: str
```

File: to_nwb/neuroscope.py [hunk 9]
```diff
@@ -318,57 +349,68 @@ def write_electrode_table(nwbfile, session_path, electrode_positions=None,
...
+    lfp_fs = get_lfp_sampling_rate(session_path)
+    n_channels = sum(len(x) for x in get_channel_groups(session_path))
 
-    all_channels_data = np.fromfile(lfp_filepath, dtype=np.int16).reshape(-1, n_channels)
+    if os.path.isfile(lfp_filepath):
+        if stub:
+            max_size = 50
+            all_channels_data = np.fromfile(lfp_filepath,
+                                            dtype=np.int16,
+                                            count=max_size*n_channels).reshape(-1, n_channels)
```
Comment (bendichter) [inline to_nwb/neuroscope.py]: oh cool, I didn't know about the count arg. This is nicely done.

File: to_nwb/neuroscope.py [hunk 10]
```diff
@@ -318,57 +349,68 @@ def write_electrode_table(nwbfile, session_path, electrode_positions=None,
...
+        print("Warning: No .eeg file found at the path location!"
+              "Unable to retrieve all_channels_data.")
+        all_channels_data = None
 
     return lfp_fs, all_channels_data
 
 
-def write_lfp(nwbfile, data, fs, name='LFP', description='local field potential signal', electrode_inds=None):
+def write_lfp(nwbfile: NWBFile, data: ArrayLike, fs: float,
+              electrode_inds: Optional[List[int]],
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
def write_lfp(nwbfile: NWBFile, data: ArrayLike, fs: float,
              electrode_inds: Optional[List[int]] = None,
```

File: to_nwb/neuroscope.py [hunk 11]
```diff
@@ -409,15 +454,15 @@ def add_lfp(nwbfile, session_path, name='LFP', description='local field potentia
...
-
     fs, data = read_lfp(session_path, stub=stub)
     shank_channels = get_shank_channels(session_path)
     all_shank_channels = np.concatenate(shank_channels)
     write_lfp(nwbfile, data[:, all_shank_channels], fs, name, description)
 
 
-def get_events(session_path, suffixes=None):
-    """
+def get_events(session_path: str, suffixes: Iterable[int]):
```
Comment (bendichter) [inline to_nwb/neuroscope.py]: nice use of Iterable

File: to_nwb/neuroscope.py [hunk 12]
```diff
@@ -542,23 +592,24 @@ def write_spike_waveforms(nwbfile, session_path, shankn, stub=False, compression
     check_module(nwbfile, 'ecephys').add_data_interface(spike_event_series)
 
 
-def add_units(nwbfile, session_path, custom_cols=None, max_shanks=8):
-    """
+def add_units(nwbfile: NWBFile, session_path: str,
+              custom_cols: Optional[List[dict]],
```
Comment (bendichter) [inline to_nwb/neuroscope.py]:
```suggestion
              custom_cols: Optional[List[dict]] = None,
```

**Review by bendichter — CHANGES_REQUESTED** [review]
looks good overall. A few requested changes.

File: to_nwb/neuroscope.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: So you would prefer if it were an `assert`?

Comment (bendichter) [reply to_nwb/neuroscope.py]: Yes, that sounds like a good idea!

**Comment by bendichter** [comment]
@CodyCBakerPhD I don't really understand why you are editing this file. Is this necessary for advancing NWB Conversion Tools?

**Comment by CodyCBakerPhD** [comment]
The changes to the stubbing methods are necessary for ensuring the `buzsaki-lab-to-nwb` converter is processing the original data correctly; this is how I detected those output differences for the LFP-related computations that are still being investigated.

The other changes are just small things to spruce it up; for the `buzsaki-lab-to-nwb`, we end up using this repository just as much as `SpikeInterface` (more, if raw data is missing), so I figured it should be maintained similarly. If you don't want any of the other proposed changes, however, just let me know and these stub changes are all we need for now.

Nothing about this directly effects `nwb-conversion-tools` unless we eventually decide to generalize some of the other DataInterface classes (such as LFPDataInterface) away from their current specific usage in `buzsaki-lab-to-nwb`.

File: to_nwb/neuroscope.py (hunk 8 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

File: to_nwb/neuroscope.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

File: to_nwb/neuroscope.py (hunk 2 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

File: to_nwb/neuroscope.py (hunk 12 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

File: to_nwb/neuroscope.py (hunk 5 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

File: to_nwb/neuroscope.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

File: to_nwb/neuroscope.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply to_nwb/neuroscope.py]: Fixed

**Comment by CodyCBakerPhD** [comment]
OK, all changes implemented. This particular update should be good to merge if you agree.

**Pull request merged**